src/NodeVisitor/CorrectnessNodeVisitor.php

```diff
@@ -0,0 +1,131 @@
+    {
+        if ($node instanceof ModuleNode) {
+            $this->rootNodes = new \SplObjectStorage();
+            $this->hasParent = $node->hasNode('parent');
+
+            // allows to identify when we enter/leave the block nodes
+            $this->blockNodes = new \SplObjectStorage();
+            foreach ($node->getNode('blocks') as $n) {
+                $this->blockNodes->attach($n);
+            }
+
+            $body = $node->getNode('body')->getNode('0');
+            // see Parser::subparse() which does not wrap the parsed Nodes if there is only one node
+            foreach (count($body) ? $body : new Node([$body]) as $k => $n) {
+                // check that this root node of a child template only contains empty output nodes
+                if ($this->hasParent && !$this->isEmptyOutputNode($n)) {
+                    throw new SyntaxError('A template that extends another one cannot include content outside Twig blocks. Did you forget to put the content inside a {% block %} tag?', $n->getTemplateLine(), $n->getSourceContext());
+                }
+                $this->rootNodes->attach($n);
+            }
+
+            return $node;
+        }
+
+        if ($this->blockNodes->contains($node)) {
+            ++$this->currentBlockNodeLevel;
+        }
+
+        if ($this->hasParent && $node->getNodeTag() && !$node instanceof BlockReferenceNode) {
+            $this->currentTagNode = $node;
+        }
+
+        if ($node instanceof ConfigNode && !$this->rootNodes->contains($node)) {
+            throw new SyntaxError(sprintf('The "%s" tag must always be at the root of the body of a template.', $node->getNodeTag()), $node->getTemplateLine(), $node->getSourceContext());
+        }
+
+        if ($this->currentTagNode && $node instanceof BlockReferenceNode) {
+            if ($this->currentTagNode instanceof NodeCaptureInterface || count($this->blockNodes) > 1) {
+                trigger_deprecation('twig/twig', '3.14', \sprintf('Having a "block" tag under a "%s" tag (line %d) is deprecated in %s at line %d.', $this->currentTagNode->getNodeTag(), $this->currentTagNode->getTemplateLine(), $node->getSourceContext()->getName(), $node->getTemplateLine()));
+            } else {
+                throw new SyntaxError(\sprintf('A "block" tag cannot be under a "%s" tag (line %d).', $this->currentTagNode->getNodeTag(), $this->currentTagNode->getTemplateLine()), $node->getTemplateLine(), $node->getSourceContext());
+            }
+        }
+
+        return $node;
+    }
+
+    public function leaveNode(Node $node, Environment $env): Node
+    {
+        if ($node instanceof ModuleNode) {
+            $this->rootNodes = null;
+            $this->hasParent = false;
+            $this->blockNodes = null;
+            $this->currentBlockNodeLevel = 0;
+        }
+        if ($this->hasParent && $node->getNodeTag() && !$node instanceof BlockReferenceNode) {
+            $this->currentTagNode = null;
```

Comment:
This looks wrong to me, as it will reset the current tag node to `null` rather than resetting it to its state before entering that node (try putting a block inside a nested `if` in a template followed by a block after the inner `if` but inside the outer one, and your current code would not report the second block as being incorrect)